### PR TITLE
[Resnet] Create resnet model @open sesame 06/30 16:02

### PR DESCRIPTION
**Pull request by zhoonit**
- [Resnet] Create resnet model

```
This patch creates code to create resnet.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: Applications/Resnet/jni/main.cpp [hunk 1]
```diff
@@ -65,23 +68,140 @@ static std::string withKey(const std::string &key,
...
+                {withKey("unit", 100), withKey("activation", "softmax")}));
+
+  return layers;
+}
+
+ModelHandle createResnet18() {
+  ModelHandle model =
+    ml::train::createModel(ml::train::ModelType::NEURAL_NET,
+                           {withKey("loss", "cross"),
+                            withKey("batch_size", 128), withKey("epochs", 60)});
```
Comment (kparichay) [inline Applications/Resnet/jni/main.cpp:180]: You might want to reduce training times by a smaller number of epochs (possibly use pre-trained resnet18 for initialization) for a later PR.

File: Applications/Resnet/jni/main.cpp [hunk 2]
```diff
@@ -65,23 +68,140 @@ static std::string withKey(const std::string &key,
...
+
+  for (auto layers : createResnet18Graph()) {
+    model->addLayer(layers);
```
Comment (kparichay) [inline Applications/Resnet/jni/main.cpp:183]: just for debug purposes, can you shuffle the layers vector and then add layers - it should result in the same model.

File: Applications/Resnet/jni/main.cpp [hunk 3]
```diff
@@ -65,23 +68,140 @@ static std::string withKey(const std::string &key,
...
+  return layers;
+}
+
+ModelHandle createResnet18() {
+  ModelHandle model =
+    ml::train::createModel(ml::train::ModelType::NEURAL_NET,
+                           {withKey("loss", "cross"),
+                            withKey("batch_size", 128), withKey("epochs", 60)});
+
+  for (auto layers : createResnet18Graph()) {
```
Comment (kparichay) [inline Applications/Resnet/jni/main.cpp:182]: another suggestion, to be able to extend to other resnet models, use width, and number of block as a configuration variable. It does not has to be done in this PR, but keep it as a TODO.

File: Applications/Resnet/jni/main.cpp [hunk 4]
```diff
@@ -65,23 +68,140 @@ static std::string withKey(const std::string &key,
...
+  }
+
+  layers.push_back(createLayer("pooling2d", {withKey("name", "last_p1"),
+                                             withKey("pooling", "average"),
+                                             withKey("pool_size", {4, 4})}));
+
+  layers.push_back(createLayer("flatten", {withKey("name", "last_f1")}));
+  layers.push_back(
+    createLayer("fully_connected",
+                {withKey("unit", 100), withKey("activation", "softmax")}));
```
Comment (kparichay) [inline Applications/Resnet/jni/main.cpp:171]: i recommend taking `100` as a label size, and move resnet model creation to `resnet_model.cpp` for modularity.

File: Applications/Resnet/jni/main.cpp [hunk 5]
```diff
@@ -65,23 +68,140 @@ static std::string withKey(const std::string &key,
...
+ * @return vector of layers that contain full graph of resnet18
  */
-ModelHandle createResnet18() {
+std::vector<LayerHandle> createResnet18Graph() {
+  using ml::train::createLayer;
+
   std::vector<LayerHandle> layers;
-  return nullptr;
+
+  layers.push_back(
```
Comment (kparichay) [inline Applications/Resnet/jni/main.cpp:135]: i wonder where is the input layer? :)

File: Applications/Resnet/jni/main.cpp (hunk 1 above)
Comment (zhoonit) [reply Applications/Resnet/jni/main.cpp:180]: yes, I'll tune epochs when training

File: Applications/Resnet/jni/main.cpp (hunk 2 above)
Comment (zhoonit) [reply Applications/Resnet/jni/main.cpp:183]: I think there is already a bug, I'll do that after verification process is done :)

File: Applications/Resnet/jni/main.cpp (hunk 3 above)
Comment (zhoonit) [reply Applications/Resnet/jni/main.cpp:182]: Yes, as we have resnet block it can be easily configurable. I"ll add it as a todo

File: Applications/Resnet/jni/main.cpp (hunk 4 above)
Comment (zhoonit) [reply Applications/Resnet/jni/main.cpp:171]: Do we have label size property? could you explain a bit more?

File: Applications/Resnet/jni/main.cpp (hunk 5 above)
Comment (zhoonit) [reply Applications/Resnet/jni/main.cpp:135]: when we have input_shape property, we don't need input layer. Could you clarify this?

File: Applications/Resnet/jni/main.cpp (hunk 4 above)
Comment (zhoonit) [reply Applications/Resnet/jni/main.cpp:171]: I'll consider moving to resnet_model.cpp when tidying up. For now main is empty so, it's less complex

File: Applications/Resnet/jni/main.cpp (hunk 2 above)
Comment (kparichay) [reply Applications/Resnet/jni/main.cpp:183]: alright.

File: Applications/Resnet/jni/main.cpp (hunk 4 above)
Comment (kparichay) [reply Applications/Resnet/jni/main.cpp:171]: > Do we have label size property? could you explain a bit more?

Its just to modularize createResnetModel() function, not for layer property. As 100 is being unit size for last layer, keep it as a property of createResnetModel() will allow to use resnet for different size of labels. 

On a side note, when tidying up, add an option to load with/without last layer.

File: Applications/Resnet/jni/main.cpp (hunk 5 above)
Comment (kparichay) [reply Applications/Resnet/jni/main.cpp:135]: Isn't input layer necessary for model creation. I maybe wrong here.

Comment (zhoonit) [reply Applications/Resnet/jni/main.cpp:136]: afaik, if input_shape is there it is treated input. I have tested it's running fine but if there is a problem, I'll look into it :)

File: Applications/Resnet/jni/main.cpp (hunk 4 above)
Comment (zhoonit) [reply Applications/Resnet/jni/main.cpp:172]: Ah, I see, I'll take care of it on upcoming PRs :).

> On a side note, when tidying up, add an option to load with/without last layer.

Could you explain for this? is it for to skip loss layer? or more like `include_top = false` in keras?

Comment (kparichay) [reply Applications/Resnet/jni/main.cpp:172]: Yes.

File: Applications/Resnet/jni/main.cpp (hunk 5 above)
Comment (kparichay) [reply Applications/Resnet/jni/main.cpp:136]: cool.

File: Applications/Resnet/jni/main.cpp (hunk 4 above)
Comment (kparichay) [reply Applications/Resnet/jni/main.cpp:172]: > I'll consider moving to resnet_model.cpp when tidying up. For now main is empty so, it's less complex

Please also consider exposing two functions with c and c++ API model data types.

**Review by kparichay — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM
